MMCore/Devices/StorageInstance.cpp

```diff
@@ -0,0 +1,254 @@
+}
+
+int StorageInstance::GetPixelType(const char* handle, MM::StorageDataType& dataType)
+{
+   RequireInitialized(__func__);
+   int ret = GetImpl()->GetDataType(handle, dataType);
+   if (ret != DEVICE_OK)
+      return ret;
+
+   return DEVICE_OK;
+}
+
+int StorageInstance::Delete(char* handle)
+{
+   RequireInitialized(__func__);
+
+   return GetImpl()->Delete(handle);
+}
+
+int StorageInstance::List(const char* path, std::vector<std::string>& listOfDatasets)
+{
+   RequireInitialized(__func__);
+   const int maxItems(5000);
+   const int maxItemLength(1024);
+   std::vector<char*> cList(maxItems, nullptr);
+   for (auto c : cList)
+   {
+      c = new char[maxItemLength];
+      memset(c, 0, maxItemLength);
+   }
+   int ret = GetImpl()->List(path, &cList[0], maxItems, maxItemLength);
+   if (ret == DEVICE_OK)
+   {
+      listOfDatasets.clear();
+
+      for (auto c : cList)
+      {
+         if (strlen(c) == 0) break;
+         listOfDatasets.push_back(std::string(c));
+      }
+   }
+
+   for (auto c : cList) delete[] c;
+
+   return ret;
+}
+
+int StorageInstance::AddImage(const char* handle, int sizeInBytes, unsigned char* pixels, std::vector<int>& coordinates, const char* imageMeta)
+{
+   RequireInitialized(__func__);
+   return GetImpl()->AddImage(handle, sizeInBytes, pixels, &coordinates[0], (int)coordinates.size(), imageMeta);
+}
+
+int StorageInstance::AppendImage(const char* handle, int sizeInBytes, unsigned char* pixels, const char* imageMeta)
+{
+   RequireInitialized(__func__);
+   return GetImpl()->AppendImage(handle, sizeInBytes, pixels, imageMeta);
+}
+
+int StorageInstance::GetSummaryMeta(const char* handle, std::string& meta)
+{
+	char* cMeta(nullptr);
+   RequireInitialized(__func__);
+   int ret = GetImpl()->GetSummaryMeta(handle, &cMeta);
+   if (ret == DEVICE_OK)
+   {
+      if (cMeta)
+         meta = cMeta;
+      GetImpl()->ReleaseStringBuffer(cMeta);
+   }
+   return ret;
+}
+
+int StorageInstance::GetImageMeta(const char* handle, const std::vector<int>& coordinates, std::string& meta)
+{
+	char* cMeta(nullptr);
+   RequireInitialized(__func__);
+   int ret = GetImpl()->GetImageMeta(handle, const_cast<int*>(&coordinates[0]), (int)coordinates.size(), &cMeta);
+   if (ret == DEVICE_OK)
+   {
+      if (cMeta)
+         meta = cMeta;
+      GetImpl()->ReleaseStringBuffer(cMeta);
+   }
+   return ret;
+}
+
+int StorageInstance::GetCustomMeta(const char* handle, const std::string& key, std::string& meta)
+{
+   char* cMeta(nullptr);
+   RequireInitialized(__func__);
+   int ret = GetImpl()->GetCustomMetadata(handle, key.c_str(), &cMeta);
+   if (ret == DEVICE_OK)
+   {
+      if (cMeta)
+         meta = cMeta;
+      GetImpl()->ReleaseStringBuffer(cMeta);
+   }
+   return ret;
+}
+
+int StorageInstance::SetCustomMeta(const char* handle, const std::string& key, const std::string& meta)
+{
+   RequireInitialized(__func__);
+   return GetImpl()->SetCustomMetadata(handle, key.c_str(), meta.c_str());
```

Comment:
Read-only datasets do not allow setting custom metadata, correct?
If so, MMCore should check.